backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/impexp/ModCntrLogImportTableSqlUpdater.java

```diff
@@ -0,0 +1,427 @@
+import static de.metas.contracts.model.I_I_ModCntr_Log.COLUMNNAME_ModCntr_InvoicingGroup_ID;
+import static de.metas.contracts.model.I_I_ModCntr_Log.COLUMNNAME_ModCntr_Log_DocumentType;
+import static de.metas.contracts.model.I_I_ModCntr_Log.COLUMNNAME_ModCntr_Module_ID;
+import static de.metas.contracts.model.I_I_ModCntr_Log.COLUMNNAME_PriceUOM;
+import static de.metas.contracts.model.I_I_ModCntr_Log.COLUMNNAME_Price_UOM_ID;
+import static de.metas.contracts.model.I_I_ModCntr_Log.COLUMNNAME_Producer_BPartner_ID;
+import static de.metas.contracts.model.I_I_ModCntr_Log.COLUMNNAME_ProductValue;
+import static de.metas.contracts.model.I_I_ModCntr_Log.COLUMNNAME_Qty;
+import static de.metas.contracts.model.I_I_ModCntr_Log.COLUMNNAME_UOMSymbol;
+import static de.metas.contracts.model.I_I_ModCntr_Log.COLUMNNAME_WarehouseName;
+
+@UtilityClass
+public class ModCntrLogImportTableSqlUpdater
+{
+	private static final Logger logger = LogManager.getLogger(ModCntrLogImportTableSqlUpdater.class);
+	private final static String targetTableName = I_I_ModCntr_Log.Table_Name;
+
+	public void updateModularImportTable(@NonNull final ImportRecordsSelection selection)
+	{
+		dbUpdateDocumentType(selection);
+		dbUpdateContract(selection);
+		dbUpdateContractModule(selection);
+
+		dbUpdateHarvestingYear(selection);
+		dbUpdateWarehouse(selection);
+
+		dbUpdateCollectionPoint(selection);
+		dbUpdateProducer(selection);
+		dbUpdateBillPartner(selection);
+
+		dbUpdateInvoicingGroup(selection);
+
+		dbUpdateProduct(selection);
+		dbUpdateUOM(selection);
+		dbUpdatePriceUOM(selection);
+		dbUpdateCurrency(selection);
+
+		// update error message
+		dbUpdateErrorMessage(selection);
+	}
+
+	private void dbUpdateContract(@NonNull final ImportRecordsSelection selection)
+	{
+		final String sqlContractId = "SELECT c." + COLUMNNAME_C_Flatrate_Term_ID
+				+ " FROM " + I_C_Flatrate_Term.Table_Name + " c"
+				+ " WHERE c." + I_C_Flatrate_Term.COLUMNNAME_DocumentNo + " = i." + COLUMNNAME_DocumentNo
+				+ " AND c." + COLUMNNAME_AD_Org_ID + " IN (i." + COLUMNNAME_AD_Org_ID + ", 0)"
+				+ " LIMIT 1 ";
+
+		final String sql = "UPDATE " + targetTableName + " i "
+				+ " SET " + COLUMNNAME_C_Flatrate_Term_ID + " = (" + sqlContractId + ")"
+				+ " WHERE i." + COLUMNNAME_I_IsImported + "<>'Y'"
+				+ " AND i." + COLUMNNAME_C_Flatrate_Term_ID + " IS NULL"
+				+ selection.toSqlWhereClause("i");
+
+		DB.executeUpdateAndThrowExceptionOnFail(sql, ITrx.TRXNAME_ThreadInherited);
+	}
+	private void dbUpdateContractModule(@NonNull final ImportRecordsSelection selection)
+	{
+		final String sqlContractModuleId = "SELECT c." + I_ModCntr_Module.COLUMNNAME_ModCntr_Module_ID
+				+ " FROM " + I_ModCntr_Module.Table_Name + " c"
+				+ " WHERE c." + I_ModCntr_Module.COLUMNNAME_Name + " = i." + COLUMNNAME_ContractModuleName
+				+ " AND c." + COLUMNNAME_AD_Org_ID + " IN (i." + COLUMNNAME_AD_Org_ID + ", 0)"
```

Comment:
This is EXTREMELY optimistic. The modules are not required to have an unique name, in fact they will be duplicated if a contract is copied to a new year. you need to also involve the product/harvest year/calendar/sotrx and ideally BP.
Similar to how it's done in `de.metas.contracts.modular.settings.ModularContractSettingsDAO#isSettingsExist`

Reply:
<details>
<summary>Final update SQL</summary>

![image](https://github.com/metasfresh/metasfresh/assets/438169/62fc68c7-8e69-4019-a802-2b01f6b37d6e)
</details>